File: audio/triangulation.cpp
#define EIGEN_USE_BLAS
#include <eigen3/Eigen/Dense>
#include <iostream>
#include <cmath>
#include <chrono>

Eigen::Vector2f trilaterate(Eigen::RowVector2f tower1, Eigen::RowVector2f tower2, Eigen::RowVector2f tower3, float r1, float r2, float r3);

int main() {
    
    // Speed of sound in meters per second
    const int speedOfSound = 343;

    //tower positions
    Eigen::RowVector2f tower1(0.0, 0.0);
    Eigen::RowVector2f tower2(1.0, 0.0);
    Eigen::RowVector2f tower3(0.0, 1.0);

    Eigen::RowVector2f soundSource(0.75, 1.0);

    float distanceToTower1 = (soundSource - tower1).norm();
    float distanceToTower2 = (soundSource - tower2).norm();
    float distanceToTower3 = (soundSource - tower3).norm();

    std::cout << "Distance to Tower 1: " << distanceToTower1 << " meters" << std::endl;
    std::cout << "Distance to Tower 1: " << distanceToTower2 << " meters" << std::endl;
    std::cout << "Distance to Tower 1: " << distanceToTower3 << " meters" << std::endl << std::endl;

    float arrivalTime_Tower1 = 0.003644;
    float arrivalTime_Tower2 = 0.003005;
    float arrivalTime_Tower3 = 0.002187;

    std::cout << "Arrival time at Tower 1: " << arrivalTime_Tower1 << " seconds" << std::endl;
    std::cout << "Arrival time at Tower 2: " << arrivalTime_Tower2 << " seconds" << std::endl;
    std::cout << "Arrival time at Tower 3: " << arrivalTime_Tower3 << " seconds" << std::endl << std::endl;

    float r1 = arrivalTime_Tower1 * speedOfSound;
    float r2 = arrivalTime_Tower2 * speedOfSound;
    float r3 = arrivalTime_Tower3 * speedOfSound;

    auto start = std::chrono::high_resolution_clock::now();

    Eigen::Vector2f estPos = trilaterate(tower1, tower2, tower3, r1, r2, r3);

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);


    std::cout << "Estimated Position of sound source :" << estPos.transpose() << " meters" << std::endl;
    std::cout << "Time taken for trilaterate function: " << duration.count() << " microseconds" << std::endl;

}

Eigen::Vector2f trilaterate(Eigen::RowVector2f tower1, Eigen::RowVector2f tower2, Eigen::RowVector2f tower3, float r1, float r2, float r3)
{

    Eigen::RowVector2f A = 2 * (tower2 - tower1);
    Eigen::RowVector2f B = 2 * (tower3 - tower1);

    float C = (float)std::pow(r1, 2.0) - (float)std::pow(r2, 2.0)- tower1.dot(tower1) + tower2.dot(tower2);
    float D = std::pow(r1, 2.0) - std::pow(r3, 2.0) - tower1.dot(tower1) + tower3.dot(tower3);

    // Solving the sys of eqs
    Eigen::Matrix2f A_mat;
    A_mat << A(0), A(1),
             B(0), B(1);

    Eigen::Vector2f b_vec(C, D);

    Eigen::Vector2f x = A_mat.colPivHouseholderQr().solve(b_vec);

    return x;

}

File: .legacy/audio/triangulation.py
import numpy as np
import matplotlib.pyplot as plt

# Speed of sound in meters per second
speed_of_sound = 343

# Positions of the towers (in meters)
tower1 = np.array([0, 0]) # mic 4
tower2 = np.array([1, 0]) # mic 2 
tower3 = np.array([0, 1]) # mic 3

# Position of the sound source (in meters)
sound_source = np.array([0, -0.5])

# Step 1: Calculate distances from the sound source to each tower
distance_to_tower1 = np.linalg.norm(sound_source - tower1)
distance_to_tower2 = np.linalg.norm(sound_source - tower2)
distance_to_tower3 = np.linalg.norm(sound_source - tower3)

print(f"Distance to Tower 1: {distance_to_tower1:.2f} meters")
print(f"Distance to Tower 2: {distance_to_tower2:.2f} meters")
print(f"Distance to Tower 3: {distance_to_tower3:.2f} meters")

# Step 2: Calculate arrival times at each tower
arrival_time_tower1 =  distance_to_tower1 / 343
arrival_time_tower2 =  distance_to_tower2 / 343
arrival_time_tower3 =  distance_to_tower3 / 343


print(f"Arrival time at Tower 1: {arrival_time_tower1:.6f} seconds")
print(f"Arrival time at Tower 2: {arrival_time_tower2:.6f} seconds")
print(f"Arrival time at Tower 3: {arrival_time_tower3:.6f} seconds")

print(f"Time lag between Mic 4 (Ref) and Mic 2: {arrival_time_tower2 - arrival_time_tower1: .6f} seconds")
print(f"Time lag between Mic 4 (Ref) and Mic 3: {arrival_time_tower3 - arrival_time_tower1: .6f} seconds")



# Step 3: Use arrival times to estimate the location of the sound source
# We'll use trilateration to estimate the position of the sound source
# Trilateration is the process of determining absolute or relative locations of points by measurement of distances.

# Function to perform trilateration
def trilaterate(tower1, tower2, tower3, r1, r2, r3):
    A = 2 * (tower2 - tower1)
    B = 2 * (tower3 - tower1)
    C = r1**2 - r2**2 - np.dot(tower1, tower1) + np.dot(tower2, tower2)
    D = r1**2 - r3**2 - np.dot(tower1, tower1) + np.dot(tower3, tower3)
    
    # Solving the system of equations
    A_matrix = np.array([A, B]).T
    b_vector = np.array([C, D])
    
    # Solving for the position of the sound source
    estimated_position = np.linalg.solve(A_matrix, b_vector)
    return estimated_position

# Estimated distances based on arrival times
r1 = arrival_time_tower1 * speed_of_sound
r2 = arrival_time_tower2 * speed_of_sound
r3 = arrival_time_tower3 * speed_of_sound

# Estimate the position of the sound source
estimated_position = trilaterate(tower1, tower2, tower3, r1, r2, r3)

print(f"Estimated position of the sound source: ({estimated_position[0]:.2f}, {estimated_position[1]:.2f}) meters")
